pallets/xyk/src/lib.rs

```diff
@@ -2322,6 +2526,82 @@ impl<T: Config> XykFunctionsTrait<T::AccountId> for Pallet<T> {
 		Ok((liquidity_asset_id, liquidity_assets_minted))
 	}
 
+	fn provide_liquidity_with_conversion(
+		sender: T::AccountId,
+		first_asset_id: Self::CurrencyId,
+		second_asset_id: Self::CurrencyId,
+		provided_asset_id: Self::CurrencyId,
+		provided_asset_amount: Self::Balance,
+		activate_minted_liquidity: bool,
+	) -> Result<(Self::CurrencyId, Self::Balance), DispatchError> {
+		// checks
+		ensure!(!provided_asset_amount.is_zero(), Error::<T>::ZeroAmount,);
+
+		let (first_reserve, second_reserve) =
+			Pallet::<T>::get_reserves(first_asset_id, second_asset_id)?;
+
+		let (reserve, other_reserve, other_asset_id) = if provided_asset_id == first_asset_id {
+			(first_reserve, second_reserve, second_asset_id)
+		} else if provided_asset_id == second_asset_id {
+			(second_reserve, first_reserve, first_asset_id)
+		} else {
+			return Err(DispatchError::from(Error::<T>::FunctionNotAvailableForThisToken))
+		};
+
+		// Ensure user has enough tokens to sell
+		<T as Config>::Currency::ensure_can_withdraw(
+			provided_asset_id.into(),
+			&sender,
+			provided_asset_amount.into(),
+			WithdrawReasons::all(),
+			// Does not fail due to earlier ensure
+			Default::default(),
+		)
+		.or(Err(Error::<T>::NotEnoughAssets))?;
+
+		// calculate sell
+		let swap_amount =
+			Pallet::<T>::calculate_balanced_sell_amount(provided_asset_amount, reserve)?;
+
+		let bought_amount = Pallet::<T>::calculate_sell_price(reserve, other_reserve, swap_amount)?;
+
+		<Self as XykFunctionsTrait<T::AccountId>>::sell_asset(
+			sender.clone(),
+			provided_asset_id,
+			other_asset_id,
+			swap_amount,
+			bought_amount,
+		)?;
+
+		let mint_amount = provided_asset_amount
+			.checked_sub(swap_amount)
+			.ok_or_else(|| DispatchError::from(Error::<T>::MathOverflow))?;
+
+		log!(
+			info,
+			"provide_liquidity_with_conversion: ({:?}, {}, {}, {}, {}) -> ({}, {})",
+			sender,
+			first_asset_id,
+			second_asset_id,
+			provided_asset_id,
+			provided_asset_amount,
+			mint_amount,
+			bought_amount
+		);
+
+		// we swap the order of the pairs to handle rounding
+		// we spend all of the Y
+		// and have some surplus amount of X that equals to the rounded part of Y
+		<Self as XykFunctionsTrait<T::AccountId>>::mint_liquidity(
+			sender,
+			other_asset_id,
+			provided_asset_id,
+			bought_amount,
+			Self::Balance::MAX,
```

Comment:
@vanderian  Is this expected? have you agreed this functionality with OAK and Gleb?  If I understood correctly this code, this specific line will try to mint All the Y token, but if the user has some ( for some reason ) Either this operation could fail (max_sec_to_first) , either this will mint tokens that the user would not expect. 
As a User I would expect that compound will only mint tokens from rewards, not my savings